File: cppsrc/mods/cmseg.cpp
#include "cmseg.h"

#include "mkb/mkb.h"

#include "mods/freecam.h"
#include "systems/log.h"
#include "systems/pref.h"
#include "utils/draw.h"
#include "utils/patch.h"
#include "utils/timerdisp.h"

namespace cmseg {

enum class State {
    Default,
    LoadMenu,
    EnterCm,
    SegActive,
    SegComplete,
};

static State s_state = State::Default;
static Seg s_seg_request;
static u32 s_start_time;
static u32 s_seg_time;

static patch::Tramp<decltype(&mkb::g_reset_cm_course)> s_reset_cm_course_tramp;

static mkb::CourseCommand* s_overwritten_entry;
static mkb::CourseCommandOpcode s_overwritten_opcode;
static s8 s_overwritten_starting_monkeys;

static u32 s_pbs[13];

// static void debug_print_course(mkb::CourseCommand *course, u32 entry_count)
//{
//     static const char *type_strs[] = {"COURSE_CMD_IF", "COURSE_CMD_THEN", "COURSE_CMD_INFO",
//     "COURSE_CMD_END"};
//
//     mkb::OSReport("Course entry count: %d\n", entry_count);
//     for (u32 i = 0; i < entry_count; i++)
//     {
//         mkb::CourseCommand &entry = course[i];
//         mkb::OSReport("%s: n = %d, v = %d\n", type_strs[entry.opcode], entry.arg, entry.value);
//     }
//     mkb::OSReport("\n");
// }

/**
 * Create a new course in an existing one by inserting a COURSE_CMD_END entry
 */
static void gen_course(mkb::CourseCommand* course, u32 start_course_stage_num, u32 stage_count) {
    s32 start_entry_idx = -1;
    s32 end_entry_idx = -1;

    u32 curr_stage_count = 0;
    for (s32 i = 0;; i++) {
        if (course[i].opcode == mkb::COURSE_CMD_INFO && course[i].type == 0) {
            curr_stage_count++;
            if (curr_stage_count == start_course_stage_num) {
                start_entry_idx = i;
            } else if (curr_stage_count == start_course_stage_num + stage_count) {
                end_entry_idx = i;
                break;
            }
        } else if (course[i].opcode == mkb::COURSE_CMD_END) {
            if (curr_stage_count == start_course_stage_num + stage_count - 1) {
                end_entry_idx = i;  // This CourseCommand is one past the end - we tack on a
                                    // COURSE_CMD_END entry ourselves
            }
            break;
        }
    }

    // Check if we found stage indices
    MOD_ASSERT(start_entry_idx > -1);
    MOD_ASSERT(end_entry_idx > -1);

    // Write new course end marker
    s_overwritten_entry = &course[end_entry_idx];
    s_overwritten_opcode = course[end_entry_idx].opcode;
    course[end_entry_idx].opcode = mkb::COURSE_CMD_END;

    s16 first_stage_id = static_cast<s16>(course[start_entry_idx].value);
    mkb::mode_info.cm_course_stage_num = start_course_stage_num;
    mkb::mode_info.cm_stage_id = first_stage_id;  // Record first stage in course
    mkb::current_cm_entry = &course[start_entry_idx + 1];
    mkb::g_another_stage_id = first_stage_id;

    // Make up "previous" stage for "current" stage
    mkb::CmStage& curr_stage = mkb::cm_player_progress[0].curr_stage;
    curr_stage.stage_course_num = start_course_stage_num - 1;
    curr_stage.stage_id = first_stage_id - 1;

    // Next stage for player is the first one we want to start on
    mkb::CmStage& next_stage = mkb::cm_player_progress[0].next_stages[0];
    next_stage.stage_course_num = start_course_stage_num;
    next_stage.stage_id = first_stage_id;
}

static void state_load_menu() {
    mkb::g_some_other_flags &=
        ~mkb::OF_GAME_PAUSED;  // Unpause the game to avoid weird darkening issues
    mkb::main_mode_request = mkb::MD_SEL;
    // Using REINIT instead of INIT seems to prevent weird game state issues, like
    // the Final Stage sprite being shown when loading a stage in story mode
    mkb::sub_mode_request = mkb::SMD_SEL_NGC_REINIT;

    // Set menu state to have chosen Main Game -> Challenge Mode
    mkb::sel_menu_info.menu_stack_ptr = 1;
    mkb::sel_menu_info.menu_stack[0] = 0;  // ??
    mkb::sel_menu_info.menu_stack[1] = 7;  // Main game
    mkb::g_focused_root_menu = 0;
    mkb::g_focused_maingame_menu = 1;

    mkb::fade_screen_to_color(
        0x100, 0, 0x1e);  // Same as what's used to fade in stages normally in challenge mode

    s_state = State::EnterCm;
}

static const mkb::ApeCharacter s_ape_charas[] = {mkb::APE_AIAI, mkb::APE_MEEMEE, mkb::APE_BABY,
                                                 mkb::APE_GONGON};

static void state_enter_cm() {
    mkb::num_players = 1;
    s_overwritten_starting_monkeys = mkb::number_of_starting_monkeys;
    mkb::number_of_starting_monkeys = 100;
    mkb::menu_start_challenge_mode();

    // Reset this timer to prevent dark pause menu when leaving an Exit Game menu. The game draws
    // the pause menu sprite at a different depth depending on this global, which can incorrectly
    // place it behind the pause menu dim quad.
    mkb::g_playpoint_msg_counter = 0;

    s_start_time = mkb::VIGetRetraceCount();
    s_state = State::SegActive;
}

static void restore_overwritten_state() {
    s_overwritten_entry->opcode = s_overwritten_opcode;  // Restore original challenge mode course
    mkb::number_of_starting_monkeys = s_overwritten_starting_monkeys;
}

static void check_exit_seg() {
    // Restore overwritten state if we exit main mode and thus thw IW
    if (mkb::main_mode != mkb::MD_GAME) {
        restore_overwritten_state();
        s_state = State::Default;
    }
}

static void state_seg_active() {
    // Set character
    if (mkb::sub_mode_request == mkb::SMD_GAME_READY_INIT) {
        Chara ch = static_cast<Chara>(pref::get(pref::U8Pref::CmChara));
        mkb::ApeCharacter real_chara;
        if (ch == Chara::Random) {
            real_chara = s_ape_charas[mkb::rand() % 4];
        } else {
            real_chara = s_ape_charas[pref::get(pref::U8Pref::CmChara)];
        }
        mkb::active_monkey_id[0] = real_chara;
    }

    // Nuke "Final Stage" sprite
    if (s_overwritten_opcode != mkb::COURSE_CMD_END) {
        for (u32 i = 0; i < mkb::sprite_pool_info.upper_bound; i++) {
            if (mkb::sprite_pool_info.status_list[i] == 0) continue;
            mkb::Sprite& sprite = mkb::sprites[i];
            if (sprite.tick_func == mkb::sprite_final_stage_tick) {
                mkb::sprite_pool_info.status_list[i] = 0;
                break;
            }
        }
    }

    s_seg_time = mkb::VIGetRetraceCount() - s_start_time;
    if (mkb::mode_info.cm_stage_id == -1 && mkb::is_stage_complete(nullptr)) {
        u32 seg = static_cast<u32>(s_seg_request);
        if (s_seg_time < s_pbs[seg]) {
            s_pbs[seg] = s_seg_time;
        }
        s_state = State::SegComplete;
    }

    check_exit_seg();
}

void state_seg_complete() {
    // If the final stage of the segment is a bonus stage, do a custom transition back to main menu
    if (mkb::mode_info.cm_stage_id == -1 &&
        mkb::mode_info.ball_mode & mkb::BALLMODE_ON_BONUS_STAGE) {
        if (mkb::sub_mode == mkb::SMD_GAME_RINGOUT_INIT ||
            mkb::sub_mode == mkb::SMD_GAME_TIMEOVER_INIT) {
            mkb::sub_mode_frame_counter += 120;
        } else if ((mkb::sub_mode == mkb::SMD_GAME_RINGOUT_MAIN ||
                    mkb::sub_mode == mkb::SMD_GAME_TIMEOVER_MAIN) &&
                   mkb::sub_mode_frame_counter == 0x3c) {
            // Same as post-goal fadeout when completing difficulty
            mkb::fade_screen_to_color(0x101, 0, 0x3d);
            mkb::g_fade_track_volume(0x3c, 2);
        } else if (mkb::sub_mode_request == mkb::SMD_GAME_READY_INIT) {
            mkb::main_mode_request = mkb::MD_SEL;
            mkb::sub_mode_request = mkb::SMD_SEL_NGC_REINIT;
        }
    }

    // If we've reached end-of-difficulty, go back to main menu
    if (mkb::sub_mode_request == mkb::SMD_GAME_EXTRA_INIT ||
        mkb::sub_mode_request == mkb::SMD_GAME_RESULT_INIT ||
        mkb::main_mode_request == mkb::MD_AUTHOR) {
        mkb::main_mode_request = mkb::MD_SEL;
        mkb::sub_mode_request = mkb::SMD_SEL_NGC_REINIT;
    }

    check_exit_seg();
}

void init_seg() {
    mkb::CourseCommand* course = nullptr;
    u32 start_course_stage_num = 0;
    mkb::mode_flags &= ~(mkb::MF_G_PLAYING_MASTER_COURSE | mkb::MF_PLAYING_EXTRA_COURSE |
                         mkb::MF_PLAYING_MASTER_NOEX_COURSE | mkb::MF_PLAYING_MASTER_EX_COURSE);
    switch (s_seg_request) {
        case Seg::Beginner1: {
            mkb::curr_difficulty = mkb::DIFF_BEGINNER;
            course = mkb::cm_courses[0];
            start_course_stage_num = 1;
            break;
        }
        case Seg::BeginnerExtra: {
            mkb::curr_difficulty = mkb::DIFF_BEGINNER;
            mkb::mode_flags |= mkb::MF_PLAYING_EXTRA_COURSE;
            course = mkb::cm_courses[3];
            start_course_stage_num = 1;
            break;
        }
        case Seg::Advanced1: {
            mkb::curr_difficulty = mkb::DIFF_ADVANCED;
            course = mkb::cm_courses[1];
            start_course_stage_num = 1;
            break;
        }
        case Seg::Advanced11: {
            mkb::curr_difficulty = mkb::DIFF_ADVANCED;
            course = mkb::cm_courses[1];
            start_course_stage_num = 11;
            break;
        }
        case Seg::Advanced21: {
            mkb::curr_difficulty = mkb::DIFF_ADVANCED;
            course = mkb::cm_courses[1];
            start_course_stage_num = 21;
            break;
        }
        case Seg::AdvancedExtra: {
            mkb::curr_difficulty = mkb::DIFF_ADVANCED;
            mkb::mode_flags |= mkb::MF_PLAYING_EXTRA_COURSE;
            course = mkb::cm_courses[4];
            start_course_stage_num = 1;
            break;
        }
        case Seg::Expert1: {
            mkb::curr_difficulty = mkb::DIFF_EXPERT;
            course = mkb::cm_courses[2];
            start_course_stage_num = 1;
            break;
        }
        case Seg::Expert11: {
            mkb::curr_difficulty = mkb::DIFF_EXPERT;
            course = mkb::cm_courses[2];
            start_course_stage_num = 11;
            break;
        }
        case Seg::Expert21: {
            mkb::curr_difficulty = mkb::DIFF_EXPERT;
            course = mkb::cm_courses[2];
            start_course_stage_num = 21;
            break;
        }
        case Seg::Expert31: {
            mkb::curr_difficulty = mkb::DIFF_EXPERT;
            course = mkb::cm_courses[2];
            start_course_stage_num = 31;
            break;
        }
        case Seg::Expert41: {
            mkb::curr_difficulty = mkb::DIFF_EXPERT;
            course = mkb::cm_courses[2];
            start_course_stage_num = 41;
            break;
        }
        case Seg::ExpertExtra: {
            mkb::curr_difficulty = mkb::DIFF_EXPERT;
            mkb::mode_flags |= mkb::MF_PLAYING_EXTRA_COURSE;
            course = mkb::cm_courses[5];
            start_course_stage_num = 1;
            break;
        }
        case Seg::Master1: {
            mkb::curr_difficulty = mkb::DIFF_EXPERT;
            mkb::mode_flags |= mkb::MF_PLAYING_EXTRA_COURSE | mkb::MF_G_PLAYING_MASTER_COURSE |
                               mkb::MF_PLAYING_MASTER_NOEX_COURSE;
            course = mkb::cm_courses[6];
            start_course_stage_num = 1;
            break;
        }
        case Seg::MasterExtra: {
            mkb::curr_difficulty = mkb::DIFF_EXPERT;
            // Magic set of flags used in Master Extra,
            // can't be bothered to reverse all of them
            mkb::mode_flags = 0x0280071D;
            course = mkb::cm_courses[7];
            start_course_stage_num = 1;
            break;
        }
    }
    gen_course(course, start_course_stage_num, 10);
}

void request_cm_seg(Seg seg) {
    s_seg_request = seg;
    if (s_state == State::SegActive || s_state == State::SegComplete) {
        restore_overwritten_state();
    }
    if (mkb::main_mode == mkb::MD_SEL) {
        s_state = State::EnterCm;  // Load challenge mode directly
    } else {
        s_state = State::LoadMenu;  // Load main menu first
    }
}

void init() {
    patch::hook_function(s_reset_cm_course_tramp, mkb::g_reset_cm_course, []() {
        s_reset_cm_course_tramp.dest();
        if (s_state == State::SegActive) init_seg();
    });

    // Set PBs to maximum time
    for (u32& s_pb : s_pbs) {
        s_pb = 0xFFFFFFFF;
    }
}

void tick() {
    if (s_state == State::LoadMenu) {
        state_load_menu();
    } else if (s_state == State::EnterCm) {
        state_enter_cm();
    } else if (s_state == State::SegActive) {
        state_seg_active();
    } else if (s_state == State::SegComplete) {
        state_seg_complete();
    }
}

void disp() {
    if (!pref::get(pref::BoolPref::CmTimer) || freecam::should_hide_hud()) return;

    if (s_state == State::SegActive || s_state == State::SegComplete) {
        u32 seg = static_cast<u32>(s_seg_request);
        mkb::GXColor color;
        if (s_state == State::SegComplete && s_pbs[seg] == s_seg_time)
            color = draw::GOLD;
        else
            color = draw::WHITE;
        timerdisp::draw_timer(static_cast<s32>(s_seg_time), "SEG:", 0, color, false);
    }
}

}  // namespace cmseg

File: crates/smb2_practice_mod/src/mods/cmseg.rs
use core::{ffi::c_long, ptr::null_mut};

use critical_section::Mutex;
use mkb::mkb;
use num_enum::TryFromPrimitive;

use crate::{
    app::with_app,
    hook,
    systems::{
        draw,
        pref::{BoolPref, Pref, U8Pref},
    },
    utils::{misc::with_mutex, timerdisp},
};

use super::freecam::Freecam;

struct Globals {
    reset_cm_course_hook: ResetCmCourseHook,
}

static GLOBALS: Mutex<Globals> = Mutex::new(Globals {
    reset_cm_course_hook: ResetCmCourseHook::new(),
});

hook!(ResetCmCourseHook => (), mkb::g_reset_cm_course, || {
    with_mutex(&GLOBALS, |cx| {
        cx.reset_cm_course_hook.call();
    });
    with_app(|cx| {
        cx.cm_seg.on_reset_cm_course();
    });
});

#[derive(Copy, Clone, PartialEq, Default)]
enum State {
    #[default]
    Default,
    LoadMenu,
    EnterCm,
    SegActive,
    SegComplete,
}

#[derive(Copy, Clone, Default)]
pub enum Seg {
    #[default]
    Beginner1,
    BeginnerExtra,
    Advanced1,
    Advanced11,
    Advanced21,
    AdvancedExtra,
    Expert1,
    Expert11,
    Expert21,
    Expert31,
    Expert41,
    ExpertExtra,
    Master1,
    MasterExtra,
}

#[derive(Copy, Clone, TryFromPrimitive)]
#[repr(u8)]
enum Chara {
    AiAi,
    MeeMee,
    Baby,
    GonGon,
    Random,
}

pub struct CmSeg {
    state: State,
    seg_request: Seg,
    start_time: u32,
    seg_time: u32,
    overwritten_course_idx: usize,
    overwritten_entry_idx: usize,
    overwritten_opcode: mkb::CourseCommandOpcode,
    overwritten_starting_monkeys: i8,
    pbs: [u32; 14],
}

impl Default for CmSeg {
    fn default() -> Self {
        with_mutex(&GLOBALS, |cx| {
            cx.reset_cm_course_hook.hook();
        });
        Self {
            state: Default::default(),
            seg_request: Default::default(),
            start_time: 0,
            seg_time: 0,
            overwritten_course_idx: 0,
            overwritten_entry_idx: 0,
            overwritten_opcode: 0,
            overwritten_starting_monkeys: 0,
            pbs: [u32::MAX; 14],
        }
    }
}

const APE_CHARAS: [mkb::ApeCharacter; 4] = [
    mkb::APE_AIAI,
    mkb::APE_MEEMEE,
    mkb::APE_BABY,
    mkb::APE_GONGON,
];

impl CmSeg {
    fn gen_course(&mut self, course_idx: usize, start_course_stage_num: u16, stage_count: u16) {
        let mut start_entry_idx = None;
        let mut end_entry_idx = None;

        let mut curr_stage_count = 0;
        unsafe {
            let course = mkb::cm_courses[course_idx];
            let mut i = 0;
            loop {
                let entry = course.add(i as usize);
                if (*entry).opcode == mkb::COURSE_CMD_INFO as mkb::CourseCommandOpcode
                    && (*entry).type_ == 0
                {
                    curr_stage_count += 1;
                    if curr_stage_count == start_course_stage_num {
                        start_entry_idx = Some(i);
                    } else if curr_stage_count == start_course_stage_num + stage_count {
                        end_entry_idx = Some(i);
                        break;
                    }
                } else if (*entry).opcode == mkb::COURSE_CMD_END as mkb::CourseCommandOpcode {
                    if curr_stage_count == start_course_stage_num + stage_count - 1 {
                        end_entry_idx = Some(i);
                    }
                    break;
                }
                i += 1;
            }

            let start_entry_idx = start_entry_idx.unwrap();
            let end_entry_idx = end_entry_idx.unwrap();

            // Write new course end marker
            self.overwritten_course_idx = course_idx;
            self.overwritten_entry_idx = end_entry_idx;
            let overwritten_entry = course.add(end_entry_idx);
            self.overwritten_opcode = (*overwritten_entry).opcode;
            (*overwritten_entry).opcode = mkb::COURSE_CMD_END as mkb::CourseCommandOpcode;

            let first_stage_id = (*course.add(start_entry_idx as usize)).value as u16;
            mkb::mode_info.cm_course_stage_num = start_course_stage_num;
            mkb::mode_info.cm_stage_id = first_stage_id as i16;
            mkb::current_cm_entry = course.add((start_entry_idx + 1) as usize);
            mkb::g_another_stage_id = first_stage_id;

            // Make up "previous" stage for "current" stage
            let curr_stage = &mut mkb::cm_player_progress[0].curr_stage;
            curr_stage.stage_course_num = (start_course_stage_num - 1) as c_long;
            curr_stage.stage_id = (first_stage_id - 1) as c_long;

            // Next stage for player is the first one we want to start on
            let next_stage = &mut mkb::cm_player_progress[0].next_stages[0];
            next_stage.stage_course_num = start_course_stage_num as c_long;
            next_stage.stage_id = first_stage_id as c_long;
        }
    }

    unsafe fn state_load_menu(&mut self) {
        mkb::g_some_other_flags &= !mkb::OF_GAME_PAUSED;
        mkb::main_mode_request = mkb::MD_SEL;
        mkb::sub_mode_request = mkb::SMD_SEL_NGC_REINIT;

        mkb::sel_menu_info.menu_stack_ptr = 1;
        mkb::sel_menu_info.menu_stack[0] = 0;
        mkb::sel_menu_info.menu_stack[1] = 7;
        mkb::g_focused_root_menu = 0;
        mkb::g_focused_maingame_menu = 1;

        mkb::fade_screen_to_color(0x100, 0, 0x1e);

        self.state = State::EnterCm;
    }

    unsafe fn state_enter_cm(&mut self) {
        mkb::num_players = 1;
        self.overwritten_starting_monkeys = mkb::number_of_starting_monkeys;
        mkb::number_of_starting_monkeys = 100;
        mkb::menu_start_challenge_mode();

        mkb::g_playpoint_msg_counter = 0;

        self.start_time = mkb::VIGetRetraceCount();
        self.state = State::SegActive;
    }

    unsafe fn restore_overwritten_state(&mut self) {
        let overwritten_entry =
            mkb::cm_courses[self.overwritten_course_idx].add(self.overwritten_entry_idx);
        (*overwritten_entry).opcode = self.overwritten_opcode;
        mkb::number_of_starting_monkeys = self.overwritten_starting_monkeys;
    }

    unsafe fn check_exit_seg(&mut self) {
        if mkb::main_mode != mkb::MD_GAME {
            self.restore_overwritten_state();
            self.state = State::Default;
        }
    }

    unsafe fn state_seg_active(&mut self, pref: &Pref) {
        if mkb::sub_mode_request == mkb::SMD_GAME_READY_INIT {
            let ch = Chara::try_from(pref.get_u8(U8Pref::CmChara)).unwrap();
            mkb::active_monkey_id[0] = match ch {
                Chara::Random => APE_CHARAS[mkb::rand() as usize % 4],
                _ => APE_CHARAS[ch as usize],
            };
        }

        if self.overwritten_opcode != mkb::COURSE_CMD_END as mkb::CourseCommandOpcode {
            for i in 0..(mkb::sprite_pool_info.upper_bound as usize) {
                if *mkb::sprite_pool_info.status_list.add(i) == 0 {
                    continue;
                }
                let sprite = &raw mut mkb::sprites[i];
                let tick_func = (*sprite).tick_func;
                if tick_func == Some(mkb::sprite_final_stage_tick) {
                    *mkb::sprite_pool_info.status_list.add(i) = 0;
                    break;
                }
            }
        }

        self.seg_time = mkb::VIGetRetraceCount() - self.start_time;
        if mkb::mode_info.cm_stage_id == -1 && mkb::is_stage_complete(null_mut()) != 0 {
            let seg = self.seg_request as u32;
            if self.seg_time < self.pbs[seg as usize] {
                self.pbs[seg as usize] = self.seg_time;
            }
            self.state = State::SegComplete;
        }

        self.check_exit_seg();
    }

    unsafe fn state_seg_complete(&mut self) {
        if mkb::mode_info.cm_stage_id == -1
            && mkb::mode_info.ball_mode & mkb::BALLMODE_ON_BONUS_STAGE != 0
        {
            if mkb::sub_mode == mkb::SMD_GAME_RINGOUT_INIT
                || mkb::sub_mode == mkb::SMD_GAME_TIMEOVER_INIT
            {
                mkb::sub_mode_frame_counter += 120;
            } else if (mkb::sub_mode == mkb::SMD_GAME_RINGOUT_MAIN
                || mkb::sub_mode == mkb::SMD_GAME_TIMEOVER_MAIN)
                && mkb::sub_mode_frame_counter == 0x3c
            {
                mkb::fade_screen_to_color(0x101, 0, 0x3d);
                mkb::g_fade_track_volume(0x3c, 2);
            } else if mkb::sub_mode_request == mkb::SMD_GAME_READY_INIT {
                mkb::main_mode_request = mkb::MD_SEL;
                mkb::sub_mode_request = mkb::SMD_SEL_NGC_REINIT;
            }
        }

        if mkb::sub_mode_request == mkb::SMD_GAME_EXTRA_INIT
            || mkb::sub_mode_request == mkb::SMD_GAME_RESULT_INIT
            || mkb::main_mode_request == mkb::MD_AUTHOR
        {
            mkb::main_mode_request = mkb::MD_SEL;
            mkb::sub_mode_request = mkb::SMD_SEL_NGC_REINIT;
        }

        self.check_exit_seg();
    }

    unsafe fn init_seg(&mut self) {
        let course_idx;
        let start_course_stage_num;

        mkb::mode_flags &= !(mkb::MF_G_PLAYING_MASTER_COURSE
            | mkb::MF_PLAYING_EXTRA_COURSE
            | mkb::MF_PLAYING_MASTER_NOEX_COURSE
            | mkb::MF_PLAYING_MASTER_EX_COURSE);

        match self.seg_request {
            Seg::Beginner1 => {
                mkb::curr_difficulty = mkb::DIFF_BEGINNER;
                course_idx = 0;
                start_course_stage_num = 1;
            }
            Seg::BeginnerExtra => {
                mkb::curr_difficulty = mkb::DIFF_BEGINNER;
                mkb::mode_flags |= mkb::MF_PLAYING_EXTRA_COURSE;
                course_idx = 3;
                start_course_stage_num = 1;
            }
            Seg::Advanced1 => {
                mkb::curr_difficulty = mkb::DIFF_ADVANCED;
                course_idx = 1;
                start_course_stage_num = 1;
            }
            Seg::Advanced11 => {
                mkb::curr_difficulty = mkb::DIFF_ADVANCED;
                course_idx = 1;
                start_course_stage_num = 11;
            }
            Seg::Advanced21 => {
                mkb::curr_difficulty = mkb::DIFF_ADVANCED;
                course_idx = 1;
                start_course_stage_num = 21;
            }
            Seg::AdvancedExtra => {
                mkb::curr_difficulty = mkb::DIFF_ADVANCED;
                mkb::mode_flags |= mkb::MF_PLAYING_EXTRA_COURSE;
                course_idx = 4;
                start_course_stage_num = 1;
            }
            Seg::Expert1 => {
                mkb::curr_difficulty = mkb::DIFF_EXPERT;
                course_idx = 2;
                start_course_stage_num = 1;
            }
            Seg::Expert11 => {
                mkb::curr_difficulty = mkb::DIFF_EXPERT;
                course_idx = 2;
                start_course_stage_num = 11;
            }
            Seg::Expert21 => {
                mkb::curr_difficulty = mkb::DIFF_EXPERT;
                course_idx = 2;
                start_course_stage_num = 21;
            }
            Seg::Expert31 => {
                mkb::curr_difficulty = mkb::DIFF_EXPERT;
                course_idx = 2;
                start_course_stage_num = 31;
            }
            Seg::Expert41 => {
                mkb::curr_difficulty = mkb::DIFF_EXPERT;
                course_idx = 2;
                start_course_stage_num = 41;
            }
            Seg::ExpertExtra => {
                mkb::curr_difficulty = mkb::DIFF_EXPERT;
                mkb::mode_flags |= mkb::MF_PLAYING_EXTRA_COURSE;
                course_idx = 5;
                start_course_stage_num = 1;
            }
            Seg::Master1 => {
                mkb::curr_difficulty = mkb::DIFF_EXPERT;
                mkb::mode_flags |= mkb::MF_PLAYING_EXTRA_COURSE
                    | mkb::MF_G_PLAYING_MASTER_COURSE
                    | mkb::MF_PLAYING_MASTER_NOEX_COURSE;
                course_idx = 6;
                start_course_stage_num = 1;
            }
            Seg::MasterExtra => {
                mkb::curr_difficulty = mkb::DIFF_EXPERT;
                mkb::mode_flags = 0x0280071D;
                course_idx = 7;
                start_course_stage_num = 1;
            }
        }
        self.gen_course(course_idx, start_course_stage_num, 10);
    }

    pub fn request_cm_seg(&mut self, seg: Seg) {
        self.seg_request = seg;
        if self.state == State::SegActive || self.state == State::SegComplete {
            unsafe {
                self.restore_overwritten_state();
            }
        }
        unsafe {
            if mkb::main_mode == mkb::MD_SEL {
                self.state = State::EnterCm;
            } else {
                self.state = State::LoadMenu;
            }
        }
    }

    pub fn tick(&mut self, pref: &Pref) {
        unsafe {
            match self.state {
                State::LoadMenu => self.state_load_menu(),
                State::EnterCm => self.state_enter_cm(),
                State::SegActive => self.state_seg_active(pref),
                State::SegComplete => self.state_seg_complete(),
                State::Default => {}
            }
        }
    }

    pub fn draw(&self, pref: &Pref, freecam: &Freecam) {
        if !pref.get_bool(BoolPref::CmTimer) || freecam.should_hide_hud(pref) {
            return;
        }

        if self.state == State::SegActive || self.state == State::SegComplete {
            let seg = self.seg_request as usize;
            let color = if self.state == State::SegComplete && self.pbs[seg] == self.seg_time {
                draw::GOLD
            } else {
                draw::WHITE
            };

            timerdisp::draw_timer(self.seg_time as i32, "SEG:", 0, color, false);
        }
    }

    pub fn on_reset_cm_course(&mut self) {
        if self.state == State::SegActive {
            unsafe {
                self.init_seg();
            }
        }
    }
}
